### DISCUSSION TITLE
Add more fields in HTTPException

### DISCUSSION THREAD

**Discussion by ezeparziale** [fastapi/fastapi#12229] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
raise HTTPException(
    status_code=status.HTTP_409_CONFLICT,
    detail={"message": "Item already exists", "errors": errors},
)
```

### Description

I need to add more fields to HTTPException, like an errors field.

What is the best approach?

Using HTTPException:

```python
raise HTTPException(
    status_code=status.HTTP_409_CONFLICT,
    detail={"message": "Item already exists", "errors": errors},
)
```

Using JSONResponse:

```python
return JSONResponse(
    status_code=status.HTTP_409_CONFLICT,
    content={"detail": "Item already exists", "errors": errors}, 
)
```

I'm not sure what impact using JSONResponse might have.

### Operating System

macOS

### FastAPI Version

0.112.2

### Pydantic Version

2.8.2

### Python Version

3.12

**Comment by ceb10n** [comment, marked as answer]
Actually, you'll not have any significant overhead using `HTTPException`, neither will lose something returning `JSONResponse`.

Your request will run through the startlette + fastapi lifecycle, wich includes an `ExceptionMiddleware`.

So in the end, your route will be running inside a try / except inside starlette's middleware:

```python
        try:
            await app(scope, receive, sender) # app is your endpoint's handler
        except Exception as exc:
            ... #
```

And if you raise an Exception, FastAPI will return a JSONResponse, as you can see in: [exception_handlers](https://github.com/fastapi/fastapi/blob/master/fastapi/exception_handlers.py)

```python
async def http_exception_handler(request: Request, exc: HTTPException) -> Response:
    headers = getattr(exc, "headers", None)
    if not is_body_allowed_for_status_code(exc.status_code):
        return Response(status_code=exc.status_code, headers=headers)
    return JSONResponse(
        {"detail": exc.detail}, status_code=exc.status_code, headers=headers
    )
```

So basically, it's about what you think is more elegant, because technically, it will be basically the same.

In my projects, I tend to prefer custom exceptions + exceptions handlers
